File: DotNet/Backend/Controllers/UserController.cs
using Microsoft.AspNetCore.Mvc;
using Bandhan.Services;
using Bandhan.Models;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Identity;
using System.Security.Claims;
using Bandhan.DTOs;

namespace BandhanApp.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class UserController : ControllerBase
    {
        private readonly UserService _userService;

        private readonly UserManager<ApplicationUser> _userManager;

        public UserController(UserService userService, UserManager<ApplicationUser> userManager)
        {
            _userService = userService;
            _userManager = userManager;
        }

        /// <summary>
        /// Get all users (Admin only).
        /// </summary>
        [Authorize(Roles = "Admin")]
        [HttpGet("all")]
        public async Task<IActionResult> GetAllUsers()
        {
            var users = await _userService.GetAllUsers();
            if (users == null || users.Count == 0)
                return NotFound(new { message = "No users found." });

            return Ok(users);
        }

        /// <summary>
        /// Get all male users.
        /// </summary>
        [Authorize]
        [HttpGet("males")]
        public async Task<IActionResult> GetAllMales()
        {
            var users = await _userService.GetAllMales();
            if (users == null || users.Count == 0)
                return NotFound(new { message = "No male users found." });

            return Ok(users);
        }

        /// <summary>
        /// Get all female users.
        /// </summary>
        [Authorize]
        [HttpGet("females")]
        public async Task<IActionResult> GetAllFemales()
        {
            var users = await _userService.GetAllFemales();
            if (users == null || users.Count == 0)
                return NotFound(new { message = "No female users found." });

            return Ok(users);
        }

        /// <summary>
        /// Get users within a specific age range.
        /// </summary>
        [Authorize]
        [HttpGet("age-range/{min}/{max}")]
        public async Task<IActionResult> GetUsersByAge(int min, int max)
        {
            if (min < 0 || max < 0 || min > max)
                return BadRequest(new { message = "Invalid age range provided." });

            var users = await _userService.GetUsersByAgeRange(min, max);
            if (users == null || users.Count == 0)
                return NotFound(new { message = "No users found in the given age range." });

            return Ok(users);
        }

        /// <summary>
        /// Get a user by their ID.
        /// </summary>
        
        [HttpGet("{id}")]
        public async Task<IActionResult> GetUserById(Guid id) // Ensure correct data type
        {
            if (id == Guid.Empty) // Validate ID
                return BadRequest(new { message = "User ID is required." });

            var user = await _userService.GetUserById(id.ToString()); // Convert Guid to string
            if (user == null)
                return NotFound(new { message = "User not found." });

            return Ok(user);
        }


        /// <summary>
        /// Get users by salary range, excluding the current user.
        /// </summary>
        [Authorize]
        [HttpGet("salary-range/{min}/{max}/{currentUserId}")]
        public async Task<IActionResult> GetUsersBySalary(int min, int max, string currentUserId)
        {
            if (min < 0 || max < 0 || min > max)
                return BadRequest(new { message = "Invalid salary range provided." });

            if (string.IsNullOrEmpty(currentUserId))
                return BadRequest(new { message = "Current user ID is required." });

            var users = await _userService.GetUsersBySalary(min, max, currentUserId);
            if (users == null || users.Count == 0)
                return NotFound(new { message = "No users found within the given salary range." });

            return Ok(users);
        }

        /// <summary>
        /// Fetches profiles of the opposite gender for the logged-in user (Dashboard).
        /// </summary>
        
        [HttpGet("dashboard")]
        public async Task<IActionResult> GetDashboardProfiles()
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId))
                return Unauthorized(new { message = "User is not authenticated." });

            var user = await _userService.GetUserById(userId);
            if (user == null)
                return NotFound(new { message = "User not found." });

            List<ApplicationUser> profiles = (List<ApplicationUser>)((user.Gender == Gender.MALE.ToString())
                ? await _userService.GetAllFemales()
                : await _userService.GetAllMales());

            return Ok(profiles);
        }

        /// <summary>
        /// Updates the logged-in user's profile.
        /// </summary>
        [Authorize]
        [HttpPut("update")]
        public async Task<IActionResult> UpdateUser([FromBody] ApplicationUser updatedUserDetails)
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId))
                return Unauthorized(new { message = "User is not authenticated." });

            var existingUser = await _userService.GetUserById(userId);
            if (existingUser == null)
                return NotFound(new { message = "User not found." });

            // Update allowed fields
            if (!string.IsNullOrEmpty(updatedUserDetails.Name)) existingUser.Name = updatedUserDetails.Name;
            if (!string.IsNullOrEmpty(updatedUserDetails.Mobile)) existingUser.Mobile = updatedUserDetails.Mobile;
            if (!string.IsNullOrEmpty(updatedUserDetails.Location)) existingUser.Location = updatedUserDetails.Location;
            if (!string.IsNullOrEmpty(updatedUserDetails.Caste)) existingUser.Caste = updatedUserDetails.Caste;
            if (!string.IsNullOrEmpty(updatedUserDetails.Occupation)) existingUser.Occupation = updatedUserDetails.Occupation;
            if (updatedUserDetails.Income > 0) existingUser.Income = updatedUserDetails.Income;
            if (!string.IsNullOrEmpty(updatedUserDetails.Photo)) existingUser.Photo = updatedUserDetails.Photo;
            if (!string.IsNullOrEmpty(updatedUserDetails.Gender)) existingUser.Gender = updatedUserDetails.Gender;
            if (!string.IsNullOrEmpty(updatedUserDetails.MaritialStatus)) existingUser.MaritialStatus = updatedUserDetails.MaritialStatus;
            if (!string.IsNullOrEmpty(updatedUserDetails.Religion)) existingUser.Religion = updatedUserDetails.Religion;

            var result = await _userManager.UpdateAsync(existingUser);
            if (!result.Succeeded)
                return BadRequest(result.Errors);

            return Ok(new { message = "User details updated successfully." });
        }

        /// <summary>
        /// Deletes the logged-in user's account.
        /// </summary>
        [Authorize]
        [HttpDelete("delete")]
        public async Task<IActionResult> DeleteUser()
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId))
                return Unauthorized(new { message = "User is not authenticated." });

            var user = await _userService.GetUserById(userId);
            if (user == null)
                return NotFound(new { message = "User not found." });

            await _userManager.DeleteAsync(user);

            return Ok(new { message = "User account deleted successfully." });
        }

        /// <summary>
        /// Get details of the logged-in user.
        /// </summary>
        [Authorize]
        [HttpGet("me")]
        public async Task<IActionResult> GetUserDetails()
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId))
                return Unauthorized(new { message = "User is not authenticated." });

            var user = await _userService.GetUserById(userId);
            if (user == null)
                return NotFound(new { message = "User not found." });

            return Ok(user);
        }

        [Authorize]
        [HttpPost("profiles/filter")]
        public async Task<IActionResult> FilterUsers([FromBody] FilterRequest filterRequest)
        {
            // Extract user identity from JWT token
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(userId))
                return Unauthorized(new { message = "User is not authenticated." });

            // Fetch the logged-in user
            var loggedInUser = await _userManager.FindByIdAsync(userId);
            if (loggedInUser == null)
                return NotFound(new { message = "User not found." });

            // Determine opposite gender
            Gender oppositeGender = loggedInUser.Gender == Gender.MALE.ToString() ? Gender.FEMALE : Gender.MALE;

            // Debugging logs
            Console.WriteLine($"Incoming Filter Request: {filterRequest}");
            Console.WriteLine($"Logged in User: {loggedInUser.Email}");
            Console.WriteLine($"Opposite Gender for Filtering: {oppositeGender}");

            // Call service method to filter users
            var filteredUsers = await _userService.FilterUsers(filterRequest, oppositeGender);

            return Ok(filteredUsers);
        }

    }
}

File: src/main/java/com/bandhan/controller/UserController.java
package com.bandhan.controller;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.bandhan.entity.Gender;
import com.bandhan.entity.MaritialStatus;
import com.bandhan.entity.Religion;
import com.bandhan.entity.User;
import com.bandhan.exception.UserException;
import com.bandhan.repository.UserRepository;
import com.bandhan.request.FilterRequest;
import com.bandhan.response.UserResponse;
import com.bandhan.service.UserService;

@RestController
@RequestMapping("user")
public class UserController {
	@Autowired
	private UserService userService;
	
	
	
	
	
	@Autowired
	private UserRepository userRepository;
	@GetMapping("/males")
	public List<User> getAllMales()
	{
		return userService.getAllMales();
		
	}
	
	@GetMapping("/females")
	public List<User> getAllFemales()
	{
		return userService.getAllFemales();
		
	}
	

@GetMapping("/dashboard") // Make sure this endpoint exists
	    public List<User> getDashboardProfiles() {
	        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
	        
	        if (authentication == null || !authentication.isAuthenticated()) {
	            throw new RuntimeException("User is not authenticated");
	        }

	        String username = authentication.getName();
	        System.out.println("Logged-in user: " + username);

	        User user = userRepository.findByEmail(username);
	        if (user == null) {
	            throw new RuntimeException("User not found");
	        }

	        List<User> profiles = (user.getGender() == Gender.MALE) ? userService.getAllFemales() : userService.getAllMales();
	        System.out.println("Returning " + profiles.size() + " profiles");

	        return profiles;
	    }

	
	//for update user profile
		@PutMapping("/update")
		public ResponseEntity<UserResponse> updateUserHandler(@RequestBody User updatedUserDetails) throws UserException {

		    // Get the currently authenticated user
		    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
		    String loggedInEmail = authentication.getName();

		    // Fetch the existing user from the database
		    User existingUser = userRepository.findByEmail(loggedInEmail);
		    if (existingUser == null) {
		        throw new UserException("User not found for the logged-in email: " + loggedInEmail);
		    }

		    // Update only the allowed fields
		    if (updatedUserDetails.getName() != null) {
		        existingUser.setName(updatedUserDetails.getName());
		    }
		    if (updatedUserDetails.getMobile() != null) {
		        existingUser.setMobile(updatedUserDetails.getMobile());
		    }
		    if (updatedUserDetails.getDateOfBirth() != null) {
		        existingUser.setDateOfBirth(updatedUserDetails.getDateOfBirth());
		    }
		    if (updatedUserDetails.getLocation() != null) {
		        existingUser.setLocation(updatedUserDetails.getLocation());
		    }
		    if (updatedUserDetails.getCaste() != null) {
		        existingUser.setCaste(updatedUserDetails.getCaste());
		    }
		    if (updatedUserDetails.getOccupation() != null) {
		        existingUser.setOccupation(updatedUserDetails.getOccupation());
		    }
		    if (updatedUserDetails.getIncome() != null) {
		        existingUser.setIncome(updatedUserDetails.getIncome());
		    }
		    if (updatedUserDetails.getPhoto() != null) {
		        existingUser.setPhoto(updatedUserDetails.getPhoto());
		    }
		    if (updatedUserDetails.getGender() != null) {
		        existingUser.setGender(updatedUserDetails.getGender());
		    }
//		    if (updatedUserDetails.getMaritialStatus() != null) {
//		        existingUser.setMaritialStatus(updatedUserDetails.getMaritialStatus());
//		    }
		    if (updatedUserDetails.getReligion() != null) {
		        existingUser.setReligion(updatedUserDetails.getReligion());
		    }

		    // Save the updated user back to the database
		    User savedUser = userRepository.save(existingUser);

		    // Prepare the response
		    
		    UserResponse userResponse=new UserResponse();
		    userResponse.setMessage("User details updated successfully");
		    return new ResponseEntity<>(userResponse, HttpStatus.OK);
		}
	
		
		// For deleting a user profile
		@DeleteMapping("/delete")
		public ResponseEntity<UserResponse> deleteUserHandler() throws UserException {

		    // Get the currently authenticated user
		    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
		    String loggedInEmail = authentication.getName();

		    // Fetch the existing user from the database
		    User existingUser = userRepository.findByEmail(loggedInEmail);
		    if (existingUser == null) {
		        throw new UserException("User not found for the logged-in email: " + loggedInEmail);
		    }

		    // Delete the user from the database
		    userRepository.delete(existingUser);

		    // Prepare the response
		    UserResponse userResponse = new UserResponse();
		    userResponse.setMessage("User account deleted successfully");

		    return new ResponseEntity<>(userResponse, HttpStatus.OK);
		}
		
		
		 
		
		
		// Get logged-in user details
	    @GetMapping("/me")
	    public ResponseEntity<?> getUserDetails() throws UserException {
	        // Get the currently authenticated user
	        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
	        String loggedInEmail = authentication.getName();  // Get email from authentication object

	        // Fetch the existing user from the database using email
	        User existingUser = userRepository.findByEmail(loggedInEmail);

	        // Check if user exists, if not throw an exception
	        if (existingUser == null) {
	            throw new UserException("User not found for the logged-in email: " + loggedInEmail);
	        }

	        // Return only the user's name in the response
	        //String userName = existingUser.getName();  // Assuming 'getName()' method exists in your User class
	        return ResponseEntity.ok(existingUser);  // Return the user's name
	    }

		

	    
		//Search user
		  @GetMapping("/search")
		    public ResponseEntity<List<User>> searchUsers(@RequestParam(required = false) String income,
		                                                  
		                                                  @RequestParam(required = false) MaritialStatus maritialStatus,
		                                                  @RequestParam(required = false) Religion religion,
		                                                  @RequestParam(required = false) String location) {
		        List<User> users = userService.searchUsers(income, maritialStatus, religion, location);
		        return ResponseEntity.ok(users);
		    }

		  @PostMapping("/profiles/filter")
		    public List<User> filterUsers(@RequestBody FilterRequest filterRequest, 
		                                  @RequestHeader("Authorization") String token) {
		        
		        // Extract user gender from JWT token
		    	  Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
			        String username = authentication.getName(); // Get username (email)
			        

			       

			        User loggedInUser =  userRepository.findByEmail(username);		                                    

		        Gender oppositeGender = (loggedInUser.getGender() == Gender.MALE) ? Gender.FEMALE : Gender.MALE;
		        //  Debugging logs in Controller
		        System.out.println(" Incoming Filter Request: " + filterRequest);
		        System.out.println(" Logged in User: " + username);
		        System.out.println(" Opposite Gender for Filtering: " + oppositeGender);

		        // Call service method with request parameters and opposite gender
		        return userService.filterUsers(filterRequest, oppositeGender);
		    }
	
		//user data on the basis of id
		    @GetMapping("/{id}")
		    public Optional<User> getUserById(@PathVariable Long id) {
		        return userRepository.findById(id);
		    }

		    @GetMapping("/gender")
		    public String getUserGender() {
		        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
		        String username = authentication.getName(); // Get username (email)
		        User user =  userRepository.findByEmail(username);
		        return user.getGender().name(); // Return gender as a string (MALE/FEMALE)
		    }
	
}
